agent/src/main/java/com/appland/appmap/process/hooks/HttpServerRequest.java

```diff
@@ -1,58 +1,98 @@
+  private static final String URI_TEMPLATE_VARIABLES_ATTRIBUTE = "org.springframework.web.servlet.HandlerMapping.uriTemplateVariables";
+  private static final String LAST_EVENT_KEY = "com.appland.appmap.lastEvent";
   private static final Recorder recorder = Recorder.getInstance();
 
   private static void recordHttpServerRequest(Event event, HttpServletRequest req) {
     if (req.getRequestURI().endsWith(RemoteRecordingManager.RecordRoute)) {
       return;
     }
 
-    recordHttpServerRequest(event,
-      req.getMethod(), req.getRequestURI(), req.getProtocol(), 
-      req.getHeaders(),
-      req.getParameterMap());
+    recordHttpServerRequest(event, req,
+        req.getMethod(), req.getRequestURI(), req.getProtocol(),
+        req.getHeaders(),
+        req.getParameterMap());
   }
 
-  public static void recordHttpServerRequest(Event event, 
-    String method, String uri, String protocol, 
-    Map<String, String> headers,
-    Map<String, String[]> params)  {
+  private static void recordHttpServerRequest(Event event, HttpServletRequest req,
+      String method, String uri, String protocol,
+      Map<String, String> headers,
+      Map<String, String[]> params) {
     event.setHttpServerRequest(method, uri, protocol, headers);
     event.setParameters(null);
-    
-    if (params != null) {
-      for (Map.Entry<String, String[]> param : params.entrySet()) {
-        final String[] values = param.getValue();
-        event.addMessageParam(param.getKey(), values.length > 0 ? values[0] : "");
-      }
+
+    for (Map.Entry<String, String[]> param : params.entrySet()) {
+      final String[] values = param.getValue();
+      event.addMessageParam(param.getKey(), values.length > 0 ? values[0] : "");
     }
+
+    event.pending(true);
+    req.setAttribute(LAST_EVENT_KEY, event);
+
+    recorder.add(event);
+  }
+
+  public static void recordHttpServerRequest(Event event,
+      String method, String uri, String protocol,
+      Map<String, String> headers) {
+    event.setHttpServerRequest(method, uri, protocol, headers);
+    event.setParameters(null);
+    event.pending(false);
+
     recorder.add(event);
   }
 
-  private static void recordHttpServerResponse(Event event, HttpServletResponse res) {
-    recordHttpServerResponse(event, res.getStatus(), res.getHeaders());
+  private static void recordHttpServerResponse(Event event, HttpServletRequest req, HttpServletResponse res) {
+    recordHttpServerResponse(event, req, res.getStatus(), res.getHeaders());
   }
 
-  public static void recordHttpServerResponse(Event event, int status, Map<String,String> headers) {
+  public static void recordHttpServerResponse(Event event, HttpServletRequest req, int status,
+      Map<String, String> headers) {
     event.setHttpServerResponse(status, headers);
     event.setParameters(null);
+
+    if (req != null) {
+      Event requestEvent = (Event) req.getAttribute(LAST_EVENT_KEY);
```

Comment:
Should we clear LAST_EVENT_KEY at this point? (ditto, below).

Reply:
Yup, will do.